gpcontrib/arenadata_toolkit/src/bloom_set.c

```diff
@@ -0,0 +1,280 @@
+/*
+ * Set of blooms. Main entry point to find a bloom and work with it.
+ * Used to track create, extend, truncate events.
+ */
+
+#include "bloom_set.h"
+
+#include <libpq/md5.h>
+
+#define BLOOM_ENTRY_GET(set, i) (void *)(set->bloom_entries + i * FULL_BLOOM_ENTRY_SIZE(set->bloom_size));
+
+static bloom_entry_t * bloom_entry_init(const uint32_t bloom_size, void *mem)
+{
+	bloom_entry_t *bloom_entry = mem;
+
+	bloom_entry->lock = LWLockAssign();
+	bloom_entry->dbid = InvalidOid;
+	(void)bloom_init(bloom_size, &bloom_entry->bloom);
+
+	return bloom_entry;
+}
+
+bloom_set_t *
+bloom_set_init(const uint32_t bloom_count, const uint32_t bloom_size, void *mem)
+{
+	bloom_set_t *bloom_set = mem;
+
+	bloom_set->bloom_count = bloom_count;
+	bloom_set->bloom_size = bloom_size;
+	bloom_set->lock = LWLockAssign();
+
+	for (uint32_t i = 0; i < bloom_count; i++)
+	{
+		void	   *bloom_entry_mem = BLOOM_ENTRY_GET(bloom_set, i);
+
+		(void)bloom_entry_init(bloom_size, bloom_entry_mem);
+	}
+
+	return bloom_set;
+}
+
+/* simple linear search, probably should be reworked (depends on target dbs count) */
+static bloom_entry_t * find_bloom_entry(bloom_set_t * bloom_set, Oid dbid, bool *found)
+{
+	bloom_entry_t *bloom_entry;
+	int			i;
+
+	*found = false;
+
+	for (i = 0; i < bloom_set->bloom_count; i++)
+	{
+		bloom_entry = BLOOM_ENTRY_GET(bloom_set, i);
+		if (bloom_entry->dbid == dbid || bloom_entry->dbid == InvalidOid)
+			break;
+	}
+
+	if (i == bloom_set->bloom_count)
+		return NULL;
+
+	if (bloom_entry->dbid != InvalidOid)
+		*found = true;
+
+	return bloom_entry;
+}
+
+/* bind not used filter to given dbid */
+bool
+bloom_set_bind(bloom_set_t * bloom_set, Oid dbid)
+{
+	bloom_entry_t *bloom_entry;
+	bool		found;
+
+	LWLockAcquire(bloom_set->lock, LW_SHARED);
+	bloom_entry = find_bloom_entry(bloom_set, dbid, &found);
+	LWLockRelease(bloom_set->lock);
+
+	if (found)
+		return true;
+	else if (!bloom_entry)
+	{
+		return false;
+	}
+
+	LWLockAcquire(bloom_set->lock, LW_EXCLUSIVE);
+	bloom_entry = find_bloom_entry(bloom_set, dbid, &found);
```

Comment:
why do we need second iteration over the bloom set? Can't we do everything with just 1 iteration?
Thus I think we can make `bloom_set_bind()` more compact and simple.

The same for `bloom_set_unbind()`